code/trans.c

```diff
@@ -0,0 +1,377 @@
+/* trans.c: TRANSFORMS IMPLEMENTATION
+ *
+ * $Id$
+ * Copyright 2011-2022 Ravenbrook Limited.  See end of file for license.
+ *
+ * A transform is a special kind of garbage collection that replaces references
+ * to a set of objects.  The transform is piggybacked onto a garbage
+ * collection by overriding the fix method for a trace.  The mapping used to
+ * replace the references is built up in a hash table by the client.
+ *
+ *
+ * Rationale
+ *
+ * This design was arrived at after some pain.  The MPS isn't really designed
+ * for this kind of thing, and the pools generally assume that they're doing
+ * a garbage collection when they're asked to condemn, scan, fix, and reclaim
+ * stuff.  This makes it very hard to apply the transform without also doing
+ * a garbage collection.  Changing this would require a significant reworking
+ * of the MPS to generalise its ideas, and would bloat the pool classes.
```

Comment:
m. This statement has not aged well, and in any case should be moved to the design. rule.generic.self, rule.code.design

Reply:
Fixed: Moved to design, updated, clarified, and referenced in ebe0af12c6